code/datums/ai/_ai_controller.dm

```diff
@@ -130,6 +132,54 @@ multiple modular subtrees with behaviors
 	RegisterSignal(pawn, COMSIG_MOB_STATCHANGE, PROC_REF(on_stat_changed))
 	RegisterSignal(pawn, COMSIG_MOB_LOGIN, PROC_REF(on_sentience_gained))
 	RegisterSignal(pawn, COMSIG_QDELETING, PROC_REF(on_pawn_qdeleted))
+	RegisterSignal(pawn, COMSIG_GRID_UPDATED, PROC_REF(update_grid))
+
+	our_cells = new(interesting_dist, interesting_dist)
+	set_new_cells()
+
+/datum/ai_controller/proc/update_grid(datum/source, datum/spatial_grid_cell/new_cell)
+	SIGNAL_HANDLER
+
+	set_new_cells()
+
+/datum/ai_controller/proc/set_new_cells()
+
+	for(var/datum/old_grid as anything in our_cells.member_cells)
+		UnregisterSignal(old_grid, list(SPATIAL_GRID_CELL_ENTERED(SPATIAL_GRID_CONTENTS_TYPE_CLIENTS), SPATIAL_GRID_CELL_EXITED(SPATIAL_GRID_CONTENTS_TYPE_CLIENTS)))
+
+	our_cells.recalculate_cells(get_turf(pawn))
+
+	var/found_member
+	for(var/datum/spatial_grid_cell/new_grid as anything in our_cells.member_cells)
+		RegisterSignal(new_grid, SPATIAL_GRID_CELL_ENTERED(SPATIAL_GRID_CONTENTS_TYPE_CLIENTS), PROC_REF(on_client_enter))
+		RegisterSignal(new_grid, SPATIAL_GRID_CELL_EXITED(SPATIAL_GRID_CONTENTS_TYPE_CLIENTS), PROC_REF(on_client_exit))
+		if(!found_member)
+			found_member = length(new_grid.client_contents)
```

Comment:
You will still need to handle the "update idle" work though. I think the best thing to do is to see if we lost any activity, and if we gained any.

Can intelligently either recalculate, do nothing, or become active based off that